File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/misc/PathPlayer.java
package de.cubbossa.pathfinder.misc;

import de.cubbossa.disposables.Disposable;
import java.util.UUID;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.ComponentLike;

public interface PathPlayer<P> extends Disposable {

    static <P> PathPlayer<P> wrap(P player) {
        return (PathPlayer<P>) PathPlayerProvider.<P>get().wrap(player);
    }

    static <P> PathPlayer<P> wrap(UUID uuid) {
        return (PathPlayer<P>) PathPlayerProvider.<P>get().wrap(uuid);
    }

    static <P> PathPlayer<P> consoleSender() {
        return (PathPlayer<P>) PathPlayerProvider.<P>get().consoleSender();
    }

    UUID getUniqueId();

    Class<P> getPlayerClass();

    String getName();

    Component getDisplayName();

    Location getLocation();

    boolean hasPermission(String permission);

    P unwrap();

    void sendMessage(ComponentLike message);
}

File: pathfinder-api/src/main/java/de/cubbossa/pathfinder/misc/PathPlayer.kt
package de.cubbossa.pathfinder.misc

import de.cubbossa.disposables.Disposable
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.ComponentLike
import java.util.*

fun <P> pathPlayer(player: P): PathPlayer<P> {
    return PathPlayerProvider.get<P>().wrap(player) as PathPlayer<P>
}

fun <P> pathPlayer(uuid: UUID?): PathPlayer<P> {
    return PathPlayerProvider.get<P>().wrap(uuid) as PathPlayer<P>
}

fun <P> pathConsoleSender(): PathPlayer<P> {
    return PathPlayerProvider.get<P>().consoleSender() as PathPlayer<P>
}

interface PathPlayer<P> : Disposable {

    val uniqueId: UUID
    val playerClass: Class<P>
    val name: String?
    val displayName: Component?
    val location: Location

    fun hasPermission(permission: String): Boolean

    fun unwrap(): P

    fun sendMessage(message: ComponentLike)
}
